framework/src/eventloop/eventloop_event.c

```diff
@@ -81,283 +79,321 @@ static event_group_t *get_event_group(int type)
+	event_group = get_event_group(event_handler->event_type);
 	if (event_group == NULL) {
-		event_group = eventloop_new_event_group(type);
+		event_group = eventloop_new_event_group(event_handler->event_type);
 		if (event_group == NULL) {
 			return ERROR;
 		}
-	}
-	event_node = (event_node_t *)EL_ALLOC(sizeof(event_node_t));
-	if (event_node == NULL) {
-		eldbg("Failed to allocate event node\n");
-		if (sq_empty(&event_group->event_list)) {
-			sq_rem((FAR sq_entry_t *)event_group, &g_event_list);
-			EL_FREE(event_group);
+		handle_ptr = (el_handle_t *)sq_peek(&event_group->event_list);
+		while (handle_ptr) {
+			if (handle_ptr->pid == handle->pid) {
+				/* Remove old event handler for type */
+				sq_rem((FAR sq_entry_t *)handle_ptr, &event_group->event_list);
+				EL_FREE(handle_ptr);
+				break;
+			}
+			handle_ptr = (el_handle_t *)sq_next(handle_ptr);
 		}
-		return ERROR;
 	}
-	event_node->flink = NULL;
-	event_node->handle = handle;
-	sq_addlast((FAR sq_entry_t *)event_node, &event_group->event_list);
+
+	sq_addlast((FAR sq_entry_t *)handle, &event_group->event_list);
+	eventloop_increase_app_handle_cnt(handle->pid);
 
 	return OK;
 }
 
-void eventloop_unregister_event_cb(el_event_t *handle)
+static int eventloop_send_event_handle(el_handle_t *handle, void *event_data, int data_size)
 {
-	event_group_t *event_group;
-	event_node_t *ptr;
-	event_data_t *data;
+	int ret;
+	bool is_empty;
+	el_handle_t *new_handle;
+	el_event_t *event;
 
-	if (handle == NULL || handle->data == NULL) {
-		return;
+	if (handle == NULL || data_size < 0) {
+		eldbg("Invalid Parameter\n");
+		return EVENTLOOP_INVALID_PARAM;
 	}
 
-	data = (event_data_t *)handle->data;
+	new_handle = (el_handle_t *)EL_ALLOC(sizeof(el_handle_t));
+	if (new_handle == NULL) {
+		return EVENTLOOP_OUT_OF_MEMORY;
+	}
+	memcpy(new_handle, handle, sizeof(el_handle_t));
+
+	event = &new_handle->data.event;
+	if (data_size == 0) {
+		event->event_data = NULL;
+	} else {
+		event->event_data = EL_ALLOC(data_size);
+		if (event->event_data == NULL) {
+			eldbg("Failed to allocate callback info\n");
+			return EVENTLOOP_OUT_OF_MEMORY;
+		}
+		memcpy(event->event_data, event_data, data_size);
+	}
 
-	event_group = get_event_group(data->type);
-	if (event_group != NULL) {
-		ptr = (event_node_t *)sq_peek(&event_group->event_list);
-		while (ptr != NULL && ptr->handle != NULL) {
-			if (ptr->handle == handle) {
-				sq_rem((FAR sq_entry_t *)ptr, &event_group->event_list);
-				if (data->event_data != NULL) {
-					EL_FREE(data->event_data);
-				}
-				EL_FREE(data);
-				EL_FREE(handle);
-				EL_FREE(ptr);
-				break;
-			}
-			ptr = (event_node_t *)sq_next(ptr);
+	ret = eventloop_job_empty(handle, &is_empty);
+	if (ret != OK) {
+		return EVENTLOOP_OPERATION_FAIL;
+	}
+
+	/* Add handle to a queue of jobs */
+	ret = eventloop_push_job(new_handle);
+	if (ret == OK && is_empty) {
+		/* Send signal to notify event */
+		ret = kill(new_handle->pid, SIGEL_EVENT);
+		if (ret == OK) {
+		} else {

```

Comment:
how about change `if` and `else` to `if (ret != OK)`

Reply:
I missed to add log message on success. I updated.